File: backend/app/routes/skills.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from ..database import get_db
from .. import schemas, crud, auth, models

router = APIRouter()


# Skill categories endpoints

@router.get("/skill-categories", response_model=List[schemas.SkillCategoryResponse])
def get_skill_categories(db: Session = Depends(get_db)):
    """
    Retrieve all skill categories with their associated skills.
    
    - Public endpoint - Accessible to all visitors
    - Returns only top-level categories (parent_id is null)
    - Each category includes its nested skills
    - Ordered by the 'order' field for custom sorting
    """
    categories = db.query(models.SkillCategory).filter(
        models.SkillCategory.parent_id == None
    ).order_by(models.SkillCategory.order).all()
    
    for category in categories:
        category.skills = crud.get_skills_by_category(db, category.id)
    
    return categories


@router.post("/admin/skill-categories", response_model=schemas.SkillCategoryResponse)
def create_skill_category(
    category: schemas.SkillCategoryCreate,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Create a new skill category.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Can be a top-level category or nested under another category
    """
    return crud.create_skill_category(db, category)


@router.put("/admin/skill-categories/{category_id}", response_model=schemas.SkillCategoryResponse)
def update_skill_category(
    category_id: int,
    category_update: schemas.SkillCategoryUpdate,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Update an existing skill category.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Returns 404 if category does not exist
    """
    category = crud.update_skill_category(db, category_id, category_update)
    if not category:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    return category


@router.delete("/admin/skill-categories/{category_id}")
def delete_skill_category(
    category_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Delete a skill category.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Cascades to delete all skills within this category
    - Returns 404 if category does not exist
    """
    if not crud.delete_skill_category(db, category_id):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    return {"message": "Category deleted successfully"}


# Skills endpoints

@router.get("/skills", response_model=List[schemas.SkillResponse])
def get_skills(db: Session = Depends(get_db)):
    """
    Retrieve all skills.
    
    - Public endpoint - Accessible to all visitors
    - Returns all skills regardless of category
    - Ordered by the 'order' field for custom sorting
    """
    return crud.get_all_skills(db)


@router.post("/admin/skills", response_model=schemas.SkillResponse)
def create_skill(
    skill: schemas.SkillCreate,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Create a new skill.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Must be associated with a skill category
    """
    return crud.create_skill(db, skill)


@router.put("/admin/skills/{skill_id}", response_model=schemas.SkillResponse)
def update_skill(
    skill_id: int,
    skill_update: schemas.SkillUpdate,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Update an existing skill.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Returns 404 if skill does not exist
    """
    skill = crud.update_skill(db, skill_id, skill_update)
    if not skill:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Skill not found"
        )
    return skill


@router.delete("/admin/skills/{skill_id}")
def delete_skill(
    skill_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Delete a skill.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Returns 404 if skill does not exist
    """
    if not crud.delete_skill(db, skill_id):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Skill not found"
        )
    return {"message": "Skill deleted successfully"}

File: frontend/src/api/skills.js
// Skills API calls - Skill categories and individual skills

import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8001';

/**
 * Create an axios instance with authentication token.
 *
 * @param {string} token - JWT access token
 * @returns {AxiosInstance} Authenticated axios instance
 */
const getApiWithToken = (token) => {
  return axios.create({
    baseURL: API_URL,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });
};

// Public endpoints - No authentication required

/**
 * Fetch all skill categories with their nested skills.
 *
 * @returns {Promise<SkillCategoryResponse[]>} List of skill categories
 */
export const getSkillCategories = async () => {
  const response = await axios.get(`${API_URL}/api/skill-categories`);
  return response.data;
};

/**
 * Fetch all skills.
 *
 * @returns {Promise<SkillResponse[]>} List of skills
 */
export const getSkills = async () => {
  const response = await axios.get(`${API_URL}/api/skills`);
  return response.data;
};

// Admin endpoints - Authentication required

// Skill Categories

/**
 * Create a new skill category.
 *
 * @param {SkillCategoryCreate} categoryData - Category data (name, icon, order, parent_id)
 * @param {string} token - JWT access token
 * @returns {Promise<SkillCategoryResponse>} Created category
 */
export const createSkillCategory = async (categoryData, token) => {
  const api = getApiWithToken(token);
  const response = await api.post('/api/admin/skill-categories', categoryData);
  return response.data;
};

/**
 * Update an existing skill category.
 *
 * @param {number} id - Category ID to update
 * @param {SkillCategoryUpdate} categoryData - Updated category data
 * @param {string} token - JWT access token
 * @returns {Promise<SkillCategoryResponse>} Updated category
 */
export const updateSkillCategory = async (id, categoryData, token) => {
  const api = getApiWithToken(token);
  const response = await api.put(`/api/admin/skill-categories/${id}`, categoryData);
  return response.data;
};

/**
 * Delete a skill category.
 *
 * @param {number} id - Category ID to delete
 * @param {string} token - JWT access token
 * @returns {Promise<{message: string}>} Deletion confirmation
 */
export const deleteSkillCategory = async (id, token) => {
  const api = getApiWithToken(token);
  const response = await api.delete(`/api/admin/skill-categories/${id}`);
  return response.data;
};

// Skills

/**
 * Create a new skill.
 *
 * @param {SkillCreate} skillData - Skill data (name, icon, level, order, skill_category_id)
 * @param {string} token - JWT access token
 * @returns {Promise<SkillResponse>} Created skill
 */
export const createSkill = async (skillData, token) => {
  const api = getApiWithToken(token);
  const response = await api.post('/api/admin/skills', skillData);
  return response.data;
};

/**
 * Update an existing skill.
 *
 * @param {number} id - Skill ID to update
 * @param {SkillUpdate} skillData - Updated skill data
 * @param {string} token - JWT access token
 * @returns {Promise<SkillResponse>} Updated skill
 */
export const updateSkill = async (id, skillData, token) => {
  const api = getApiWithToken(token);
  const response = await api.put(`/api/admin/skills/${id}`, skillData);
  return response.data;
};

/**
 * Delete a skill.
 *
 * @param {number} id - Skill ID to delete
 * @param {string} token - JWT access token
 * @returns {Promise<{message: string}>} Deletion confirmation
 */
export const deleteSkill = async (id, token) => {
  const api = getApiWithToken(token);
  const response = await api.delete(`/api/admin/skills/${id}`);
  return response.data;
};